torch/_prims/__init__.py

```diff
@@ -0,0 +1,992 @@
+import torch
+from torch._C import _add_docstr  # type: ignore[attr-defined]
+
+import torch._prims.utils as utils
+from torch._prims.utils import TensorLike, TensorMeta
+
+from typing import Sequence, Optional, Union, Callable
+from numbers import Number
+from functools import reduce
+from enum import Enum
+
+# Experimental module containing prototype "primitive" operations.
+
+Tensor = torch.Tensor
+
+__all__ = [
+    #
+    # Common datastructures and helpers
+    #
+    "RETURN_TYPE",
+    #
+    # Elementwise unary prims
+    #
+    "abs",
+    "acos",
+    "acosh",
+    "asin",
+    "atan",
+    "cos",
+    "cosh",
+    "bessel_i0e",
+    "bessel_i1e",
+    "cbrt",
+    "ceil",
+    "digamma",
+    "erf",
+    "erf_inv",
+    "erfc",
+    "exp",
+    "expm1",
+    "floor",
+    "igamma",
+    "igammac",
+    "is_finite",
+    "lgamma",
+    "log",
+    "log1p",
+    "neg",
+    "reciprocal",
+    "round",
+    "sign",
+    "sin",
+    "sinh",
+    "sqrt",
+    "square",
+    "tan",
+    #
+    # Elementwise binary prims
+    #
+    "add",
+    "atan2",
+    "bitwise_and",
+    "bitwise_not",
+    "bitwise_or",
+    "bitwise_xor",
+    # 'complex',  # needs custom meta
+    "div",
+    "eq",
+    "ge",
+    "gt",
+    "le",
+    "lt",
+    "max",
+    "min",
+    "mul",
+    "ne",
+    "nextafter",
+    "pow",
+    "rsqrt",
+    "shift_left",
+    "shift_right_arithmetic",
+    "shift_right_logical",  # not implemented
+    #
+    # View prims
+    #
+    "broadcast_in_dim",
+    "collapse_view",
+    "split_dim",
+    "squeeze",
+    #
+    # Shape prims
+    #
+    "collapse",
+    "concatenate",
+    "reshape",
+    #
+    # Conditional prims
+    #
+    "select",
+    #
+    # Data conversion and movement prims
+    #
+    "convert_element_type",
+    "device_put",
+    #
+    # Inplace prims
+    #
+    "copy_to",
+    "resize",
+]
+
+#
+# Common datastructures and helpers
+#
+
+# Describes the return type of the primitive:
+#
+#   - NEW, a new tensor is created
+#   - VIEW, a view of an input tensor is returned
+#   - INPLACE, one or more input tensors is modified
+#
+# these descriptors are mututally exclusive and exhaustive.
+class RETURN_TYPE(Enum):
+    NEW = (0,)
+    VIEW = (1,)
+    INPLACE = (2,)
+
+
+def _make_prim(
+    *, meta: Callable, impl_aten: Callable, return_type: RETURN_TYPE, doc: str
+):
+    """
+    Creates a primitive operation.
+
+    """
+
+    def _prim(*args, **kwargs):
+        # TODO: refactor this to not incur such a latency hit
+        return_meta = False
+        for arg in args:
+            if isinstance(arg, TensorMeta):
+                return_meta = True
+        for v in kwargs.values():
+            if isinstance(v, TensorMeta):
+                return_meta = True
```

Comment:
it seems a bit goofy to have a miniature implementation of the dispatcher here, but *only* for meta tensors... could have, for example, reused `handle_torch_function`, and then you would have gotten `__torch_function__` overloading for free

Reply:
It is goofy and we should fix it